File: solutions/python3/0030.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        ans = []
        dict = collections.defaultdict(int)
        for word in words:
            dict[word] += 1

        wordLen = len(words[0])
        for i in range(wordLen):
            index = i
            count = 0
            tempMap = collections.defaultdict(int)

            for j in range(i, len(s) - wordLen + 1, wordLen):
                str = s[j:j + wordLen]
                if str in dict.keys():
                    tempMap[str] += 1
                    count += 1
                    while tempMap[str] > dict[str]:
                        tempMap[s[index:index + wordLen]] -= 1
                        count -= 1
                        index += wordLen
                    if count == len(words):
                        ans.append(index)
                        tempMap[s[index:index + wordLen]] -= 1
                        count -= 1
                        index += wordLen
                else:
                    tempMap.clear()
                    count = 0
                    index = j + wordLen

        return ans

File: solutions/cpp/0030.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& words) {
        if (s.empty() || words.empty()) return {};

        vector<int> ans;
        unordered_map<string, int> map;
        for (const auto& word : words) map[word]++;

        const int m = s.size();
        const int wordLen = words[0].size();
        for (int i = 0; i < wordLen; i++) {
            int index = i;
            int count = 0;
            unordered_map<string, int> tempMap;

            for (int j = i; j <= m - wordLen; j += wordLen) {
                string str = s.substr(j, wordLen);
                if (map.count(str)) {
                    tempMap[str]++;
                    count++;
                    while (tempMap[str] > map[str]) {
                        tempMap[s.substr(index, wordLen)]--;
                        count--;
                        index += wordLen;
                    }
                    if (count == words.size()) {
                        ans.push_back(index);
                        tempMap[s.substr(index, wordLen)]--;
                        count--;
                        index += wordLen;
                    }
                } else {
                    tempMap.clear();
                    count = 0;
                    index = j + wordLen;
                }
            }
        }

        return ans;
    }
};